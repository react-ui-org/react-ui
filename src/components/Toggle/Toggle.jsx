import PropTypes from 'prop-types';
import React, { useContext } from 'react';
import { withGlobalProps } from '../../provider';
import { classNames } from '../../utils/classNames';
import { getRootValidationStateClassName } from '../_helpers/getRootValidationStateClassName';
import { transferProps } from '../_helpers/transferProps';
import { FormLayoutContext } from '../FormLayout';
import styles from './Toggle.scss';

export const Toggle = React.forwardRef((props, ref) => {
  const {
    disabled,
    helpText,
    id,
    isLabelVisible,
    label,
    labelPosition,
    required,
    validationState,
    validationText,
    ...restProps
  } = props;

  const context = useContext(FormLayoutContext);

  return (
    <label
      className={classNames(
        styles.root,
        context && styles.isRootInFormLayout,
        context && context.layout === 'horizontal' ? styles.isRootLayoutHorizontal : styles.isRootLayoutVertical,
        labelPosition === 'before' && styles.hasRootLabelBefore,
        disabled && styles.isRootDisabled,
        required && styles.isRootRequired,
        getRootValidationStateClassName(validationState, styles),
      )}
      htmlFor={id}
      id={id && `${id}__label`}
    >
      <div className={styles.field}>
        <div
          className={classNames(
            styles.label,
            !isLabelVisible && styles.isLabelHidden,
          )}
          id={id && `${id}__labelText`}
        >
          {label}
        </div>
        <input
          {...transferProps(restProps)}
          className={styles.input}
          disabled={disabled}
          id={id}
          name={id}
          ref={ref}
          required={required}
          type="checkbox"
        />
      </div>
      {helpText && (
        <div
          className={styles.helpText}
          id={id && `${id}__helpText`}
        >
          {helpText}
        </div>
      )}
      {validationText && (
        <div
          className={styles.validationText}
          id={id && `${id}__validationText`}
        >
          {validationText}
        </div>
      )}
    </label>
  );
});

Toggle.defaultProps = {
  disabled: false,
  helpText: null,
  id: undefined,
  isLabelVisible: true,
  labelPosition: 'after',
  required: false,
  validationState: null,
  validationText: null,
};

Toggle.propTypes = {
  /**
   * If `true`, the input will be disabled.
   */
  disabled: PropTypes.bool,
  /**
   * Optional help text.
   */
  helpText: PropTypes.node,
  /**
    * ID of the input HTML element. It also serves as a prefix for nested elements:
    * * `<ID>__label`
    * * `<ID>__labelText`
    * * `<ID>__helpText`
    * * `<ID>__validationText`
    */
  id: PropTypes.string,
  /**
   * If `false`, the label will be visually hidden (but remains accessible by assistive
   * technologies).
   */
  isLabelVisible: PropTypes.bool,
  /**
   * Toggle label.
   */
  label: PropTypes.string.isRequired,
  /**
   * Placement of the label relative to the input.
   */
  labelPosition: PropTypes.oneOf(['before', 'after']),
  /**
   * If `true`, the input will be required.
   */
  required: PropTypes.bool,
  /**
   * Alter the field to provide feedback based on validation result.
   */
  validationState: PropTypes.oneOf(['invalid', 'valid', 'warning']),
  /**
   * Validation message to be displayed.
   */
  validationText: PropTypes.node,
};

export const ToggleWithGlobalProps = withGlobalProps(Toggle, 'Toggle');

export default ToggleWithGlobalProps;
