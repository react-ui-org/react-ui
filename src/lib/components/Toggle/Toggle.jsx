import PropTypes from 'prop-types';
import React, { useContext } from 'react';
import { withProviderContext } from '../../provider';
import { classNames } from '../../utils/classNames';
import { getRootValidationStateClassName } from '../_helpers/getRootValidationStateClassName';
import { transferProps } from '../_helpers/transferProps';
import { FormLayoutContext } from '../FormLayout';
import withForwardedRef from '../withForwardedRef';
import styles from './Toggle.scss';

export const Toggle = ({
  checked,
  disabled,
  forwardedRef,
  helpText,
  id,
  isLabelVisible,
  label,
  labelPosition,
  required,
  validationState,
  validationText,
  value,
  ...restProps
}) => {
  const context = useContext(FormLayoutContext);

  return (
    <label
      className={classNames(
        styles.root,
        context && styles.isRootInFormLayout,
        context && context.layout === 'horizontal' ? styles.rootLayoutHorizontal : styles.rootLayoutVertical,
        labelPosition === 'before' && styles.hasRootLabelBefore,
        disabled && styles.isRootDisabled,
        required && styles.isRootRequired,
        getRootValidationStateClassName(validationState, styles),
      )}
      htmlFor={id}
      id={id && `${id}__label`}
    >
      <div className={styles.field}>
        <input
          {...transferProps(restProps)}
          checked={checked}
          className={styles.input}
          disabled={disabled}
          id={id}
          name={id}
          ref={forwardedRef}
          required={required}
          type="checkbox"
          value={value}
        />
        <div
          className={classNames(
            styles.label,
            !isLabelVisible && styles.isLabelHidden,
          )}
          id={id && `${id}__labelText`}
        >
          {label}
        </div>
      </div>
      {helpText && (
        <div
          className={styles.helpText}
          id={id && `${id}__helpText`}
        >
          {helpText}
        </div>
      )}
      {validationText && (
        <div
          className={styles.validationText}
          id={id && `${id}__validationText`}
        >
          {validationText}
        </div>
      )}
    </label>
  );
};

Toggle.defaultProps = {
  checked: undefined,
  disabled: false,
  forwardedRef: undefined,
  helpText: null,
  id: undefined,
  isLabelVisible: true,
  labelPosition: 'after',
  required: false,
  validationState: null,
  validationText: null,
  value: undefined,
};

Toggle.propTypes = {
  /**
   * If `true`, the input will be checked.
   */
  checked: PropTypes.bool,
  /**
   * If `true`, the input will be disabled.
   */
  disabled: PropTypes.bool,
  /**
   * Reference forwarded to the `input` element.
   */
  forwardedRef: PropTypes.oneOfType([
    PropTypes.func,
    // eslint-disable-next-line react/forbid-prop-types
    PropTypes.shape({ current: PropTypes.any }),
  ]),
  /**
   * Optional help text.
   */
  helpText: PropTypes.node,
  /**
    * ID of the input HTML element. It also serves as a prefix for nested elements:
    * * `<ID>__label`
    * * `<ID>__labelText`
    * * `<ID>__helpText`
    * * `<ID>__validationText`
    */
  id: PropTypes.string,
  /**
   * If `false`, the label will be visually hidden (but remains accessible by assistive
   * technologies).
   */
  isLabelVisible: PropTypes.bool,
  /**
   * Toggle label.
   */
  label: PropTypes.string.isRequired,
  /**
   * Placement of the label relative to the input.
   */
  labelPosition: PropTypes.oneOf(['before', 'after']),
  /**
   * If `true`, the input will be required.
   */
  required: PropTypes.bool,
  /**
   * Alter the field to provide feedback based on validation result.
   */
  validationState: PropTypes.oneOf(['invalid', 'valid', 'warning']),
  /**
   * Validation message to be displayed.
   */
  validationText: PropTypes.node,
  /**
   * Value of the input.
   */
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
};

export const ToggleWithContext = withForwardedRef(withProviderContext(Toggle, 'Toggle'));

export default ToggleWithContext;
