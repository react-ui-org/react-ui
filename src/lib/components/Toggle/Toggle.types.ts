export interface ToggleProps {
  /**
   * Rest of the props.
   */
  [key: string]: unknown;
  /**
   * If `true`, the input will be disabled.
   */
  disabled?: boolean;
  /**
   * Optional help text.
   */
  helpText?: React.ReactNode;
  /**
   * ID of the input HTML element. It also serves as a prefix for nested elements:
   * * `<ID>__label`
   * * `<ID>__labelText`
   * * `<ID>__helpText`
   * * `<ID>__validationText`
   */
  id?: string;
  /**
   * If `false`, the label will be visually hidden (but remains accessible by assistive
   * technologies).
   */
  isLabelVisible?: boolean;
  /**
   * Toggle label.
   */
  label: string;
  /**
   * Placement of the label relative to the input.
   */
  labelPosition?: Position;
  /**
   * If `true`, the input will be required.
   */
  required?: boolean;
  /**
   * Alter the field to provide feedback based on validation result.
   */
  validationState?: Validation;
  /**
   * Validation message to be displayed.
   */
  validationText?: React.ReactNode;
}
