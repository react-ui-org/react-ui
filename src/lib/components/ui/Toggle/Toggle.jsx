import PropTypes from 'prop-types';
import React from 'react';
import getRootValidationStateClassName from '../../../helpers/getRootValidationStateClassName';
import { withProviderContext } from '../../../provider';
import transferProps from '../../../utils/transferProps';
import withForwardedRef from '../withForwardedRef';
import styles from './Toggle.scss';

export const Toggle = ({
  changeHandler,
  checked,
  disabled,
  forwardedRef,
  helpText,
  id,
  inFormLayout,
  isLabelVisible,
  label,
  labelPosition,
  layout,
  required,
  validationState,
  validationText,
  value,
  ...restProps
}) => (
  <label
    className={[
      styles.root,
      inFormLayout ? styles.isRootInFormLayout : '',
      labelPosition === 'before' ? styles.hasRootLabelBefore : '',
      layout === 'horizontal' ? styles.rootLayoutHorizontal : styles.rootLayoutVertical,
      disabled ? styles.isRootDisabled : '',
      required ? styles.isRootRequired : '',
      getRootValidationStateClassName(validationState, styles),
    ].join(' ')}
    htmlFor={id}
    id={id && `${id}__label`}
  >
    <div className={styles.field}>
      <input
        {...transferProps(restProps)}
        checked={checked}
        className={styles.input}
        disabled={disabled}
        id={id}
        name={id}
        onChange={changeHandler}
        ref={forwardedRef}
        required={required}
        type="checkbox"
        value={value}
      />
      <div
        className={[
          styles.label,
          isLabelVisible ? '' : styles.isLabelHidden,
        ].join(' ')}
        id={id && `${id}__labelText`}
      >
        {label}
      </div>
    </div>
    {helpText && (
      <div
        className={styles.helpText}
        id={id && `${id}__helpText`}
      >
        {helpText}
      </div>
    )}
    {validationText && (
      <div
        className={styles.validationText}
        id={id && `${id}__validationText`}
      >
        {validationText}
      </div>
    )}
  </label>
);

Toggle.defaultProps = {
  changeHandler: null,
  checked: false,
  disabled: false,
  forwardedRef: undefined,
  helpText: null,
  id: undefined,
  inFormLayout: false,
  isLabelVisible: true,
  labelPosition: 'after',
  layout: 'vertical',
  required: false,
  validationState: null,
  validationText: null,
  value: undefined,
};

Toggle.propTypes = {
  /**
   * Function to call when the input is toggled.
   */
  changeHandler: PropTypes.func,
  /**
   * If `true`, the input will be checked.
   */
  checked: PropTypes.bool,
  /**
   * If `true`, the input will be disabled.
   */
  disabled: PropTypes.bool,
  /**
   * Reference forwarded to the `input` element.
   */
  forwardedRef: PropTypes.oneOfType([
    PropTypes.func,
    // eslint-disable-next-line react/forbid-prop-types
    PropTypes.shape({ current: PropTypes.any }),
  ]),
  /**
   * Optional help text.
   */
  helpText: PropTypes.node,
  /**
   * ID of the input HTML element. It also serves as a prefix for important inner elements:
   * `<ID>__label`, `<ID>__labelText`, `<ID>__helpText`, and `<ID>__validationText`.
   */
  id: PropTypes.string,
  /**
   * Treat the field differently when it's inside a FormLayout. Do not set manually!
   */
  inFormLayout: PropTypes.bool,
  /**
   * If `false`, the label will be visually hidden (but remains accessible by assistive
   * technologies).
   */
  isLabelVisible: PropTypes.bool,
  /**
   * Toggle label.
   */
  label: PropTypes.string.isRequired,
  /**
   * Placement of the label relative to the input.
   */
  labelPosition: PropTypes.oneOf(['before', 'after']),
  /**
   * Layout of the field. It has impact only on Toggle inside a FormLayout.
   */
  layout: PropTypes.oneOf(['horizontal', 'vertical']),
  /**
   * If `true`, the input will be required.
   */
  required: PropTypes.bool,
  /**
   * Alter the field to provide feedback based on validation result.
   */
  validationState: PropTypes.oneOf(['invalid', 'valid', 'warning']),
  /**
   * Validation message to be displayed.
   */
  validationText: PropTypes.node,
  /**
   * Value of the input.
   */
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
};

export const ToggleWithContext = withForwardedRef(withProviderContext(Toggle, 'Toggle'));

export default ToggleWithContext;
