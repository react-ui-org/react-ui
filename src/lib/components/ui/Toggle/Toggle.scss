@use '../../../styles/tools/form-fields/foundation';
@use '../../../styles/tools/form-fields/inline-field-elements';
@use '../../../styles/tools/form-fields/inline-field-layout';
@use '../../../styles/tools/form-fields/relationships';
@use '../../../styles/tools/form-fields/variants';
@use '../../../styles/tools/accessibility';

// Foundation
.root {
  @include foundation.root();
  @include inline-field-layout.root();
  @include variants.visual(check);
  @include relationships.horizontal-neighbor();
}

.label {
  @include foundation.label();
}

.field {
  @include inline-field-layout.field();
}

.input {
  @include inline-field-elements.check-input($type: toggle);
}

.helpText,
.validationText {
  @include foundation.help-text();
}

.isRootRequired .label {
  @include foundation.label-required();
}

// States
.isRootStateInvalid {
  @include variants.validation(invalid);
}

.isRootStateValid {
  @include variants.validation(valid);
}

.isRootStateWarning {
  @include variants.validation(warning);
}

// Invisible label
.isLabelHidden {
  @include accessibility.hide-text();
}

// Layouts
.hasRootLabelBefore {
  @include inline-field-layout.has-label-before();
}

.isRootInFormLayout {
  @include inline-field-layout.in-form-layout();
}
