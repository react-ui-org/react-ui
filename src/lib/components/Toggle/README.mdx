---
name: Toggle
menu: 'Inputs'
route: /components/toggle
---

# Toggle

Toggle switches things on and off.

import {
  Playground,
  Props,
} from 'docz'
import { Toggle } from './Toggle'

## Basic Usage

To implement the Toggle component, you need to import it first:

```js
import { Toggle } from '@react-ui-org/react-ui';
```

And use it:

<Playground>
  {() => {
    const [studioQuality, setStudioQuality] = React.useState(true);
    return (
      <Toggle
        checked={studioQuality}
        label="Listen in studio quality"
        onChange={() => setStudioQuality(!studioQuality)}
      />
    );
  }}
</Playground>

See [API](#api) for all available options.

## General Guidelines

- Use the toggle only for boolean (true/false) input **with an immediate effect
  (without confirmation)**. To toggle things on or off in forms that require to
  be submitted by a button, use rather the
  [CheckboxField](/components/checkbox-field) component.

- **Use positive wording for the toggle label,** so that it's clear what will
  happen when the user turns on the toggle. Avoid negations such as “Don't send
  me any emails” which would mean that the user needs to turn the toggle **on**
  in order for something **not** to happen.

- Only make the Toggle's label invisible when there is **another visual clue**
  to guide users through using the input.

- When a short label is not enough, use **help texts to guide users** before
  they enter anything.

- Use **clear, calm error messages** when there's a problem with what they
  entered.

- The toggle is designed for **switching things on and off.** Don't use it to
  switch between two different things that cannot be described as on or off
  using a single label, e.g. different viewing modes. In such cases, consider
  using the [ButtonGroup](/components/button-group) component.

## Help Text

You may provide an additional help text to clarify what will be the impact of
turning the toggle on or off.

<Playground>
  {() => {
    const [studioQuality, setStudioQuality] = React.useState(true);
    return (
      <Toggle
        checked={studioQuality}
        helpText={
          'Unrivaled audio quality. Uses the MQA audio format. '
          + 'Transfers a lot of data.'
        }
        label="Listen in studio quality"
        onChange={() => setStudioQuality(!studioQuality)}
      />
    );
  }}
</Playground>

## Label Options

In some cases, it may be convenient to visually hide the toggle label. The label
remains accessible to assistive technologies.

<Playground>
  {() => {
    const [studioQuality, setStudioQuality] = React.useState(true);
    return (
      <Toggle
        checked={studioQuality}
        isLabelVisible={false}
        label="You cannot see this"
        onChange={() => setStudioQuality(!studioQuality)}
      />
    );
  }}
</Playground>

It's also possible to display the label before the input:

<Playground>
  {() => {
    const [studioQuality, setStudioQuality] = React.useState(true);
    return (
      <Toggle
        checked={studioQuality}
        label="Listen in studio quality"
        labelPosition="before"
        onChange={() => setStudioQuality(!studioQuality)}
      />
    );
  }}
</Playground>

## States

### Validation States

Validation states visually present the result of validation of the input. You
should always **provide a validation message for states other than valid** so
users know what happened and what action they should take or what options they
have.

<Playground>
  {() => {
    const [studioQuality, setStudioQuality] = React.useState(true);
    return (
      <>
        <Toggle
          checked={studioQuality}
          label="Listen in studio quality"
          onChange={() => setStudioQuality(!studioQuality)}
          validationState="valid"
        />
        <Toggle
          checked={studioQuality}
          label="Listen in studio quality"
          onChange={() => setStudioQuality(!studioQuality)}
          validationState="warning"
          validationText={
            'Looks like you are connected over cellular network. '
            + 'Please make sure you know what you are doing.'
          }
        />
        <Toggle
          checked={studioQuality}
          label="Listen in studio quality"
          onChange={() => setStudioQuality(!studioQuality)}
          validationState="invalid"
          validationText="Please upgrade your plan to make this option available."
        />
      </>
    );
  }}
</Playground>

### Disabled State

Disabled state makes the input unavailable.

<Playground>
  <Toggle label="Disabled toggle" disabled />
  <Toggle
    label="Disabled toggle, checked"
    disabled
    checked
  />
</Playground>

## API

In addition to the options below, you can specify [React synthetic events] or
any custom HTML attributes that do not interfere with the API, and they will be
passed to the `input` HTML element. This enables making the component
interactive and helps improve its accessibility.

<Props table of={Toggle} />

## Theming

Head to [Forms Theming](/customize/theming/forms) to see shared form theming
options. On top of that, the following options are available for Toggle.

| Custom Property                                                    | Description                                    |
|--------------------------------------------------------------------|------------------------------------------------|
| `--rui-FormField--check__input--toggle__width`             | Input width (height is shared with other check fields) |
| `--rui-FormField--check__input--toggle__border-radius`             | Input corner radius                            |
| `--rui-FormField--check__input--toggle__background-size`           | Input background size                          |
| `--rui-FormField--check__input--toggle--default__background-image` | Background image of unchecked input            |
| `--rui-FormField--check__input--toggle--default__background-position` | Background position of unchecked input      |
| `--rui-FormField--check__input--toggle--checked__background-image` | Background image of checked input              |
| `--rui-FormField--check__input--toggle--checked__background-position` | Background position of checked input        |

[React synthetic events]: https://reactjs.org/docs/events.html
