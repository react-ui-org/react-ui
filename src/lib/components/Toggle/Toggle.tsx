import React, { useContext } from 'react';
import { withGlobalProps } from '../../provider';
import { classNames } from '../../utils/classNames';
import { getRootValidationStateClassName } from '../_helpers/getRootValidationStateClassName';
import { transferProps } from '../_helpers/transferProps';
import { FormLayoutContext } from '../FormLayout';
import styles from './Toggle.scss';
import { ToggleProps } from './Toggle.types';

export const Toggle: React.FunctionComponent<ToggleProps> = React.forwardRef<HTMLInputElement, ToggleProps>(({
  disabled = false,
  helpText,
  id,
  isLabelVisible = true,
  label,
  labelPosition = 'after',
  required = false,
  validationState,
  validationText,
  ...restProps
}, ref) => {
  const context = useContext(FormLayoutContext);

  return (
    <label
      className={classNames(
        styles.root,
        context && styles.isRootInFormLayout,
        context && context.layout === 'horizontal' ? styles.isRootLayoutHorizontal : styles.isRootLayoutVertical,
        labelPosition === 'before' && styles.hasRootLabelBefore,
        disabled && styles.isRootDisabled,
        required && styles.isRootRequired,
        getRootValidationStateClassName(validationState, styles),
      )}
      htmlFor={id}
      id={id && `${id}__label`}
    >
      <div className={styles.field}>
        <div
          className={classNames(
            styles.label,
            !isLabelVisible && styles.isLabelHidden,
          )}
          id={id && `${id}__labelText`}
        >
          {label}
        </div>
        <input
          {...transferProps(restProps)}
          className={styles.input}
          disabled={disabled}
          id={id}
          name={id}
          ref={ref}
          required={required}
          type="checkbox"
        />
      </div>
      {helpText && (
        <div
          className={styles.helpText}
          id={id && `${id}__helpText`}
        >
          {helpText}
        </div>
      )}
      {validationText && (
        <div
          className={styles.validationText}
          id={id && `${id}__validationText`}
        >
          {validationText}
        </div>
      )}
    </label>
  );
});

export const ToggleWithGlobalProps = withGlobalProps(Toggle, 'Toggle');

export default ToggleWithGlobalProps;
