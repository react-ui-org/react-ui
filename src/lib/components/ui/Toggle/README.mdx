---
name: Toggle
menu: 'UI'
route: /components/ui/toggle
---

# Toggle

Toggle switches things on and off.

import {
  Playground,
  Props,
} from 'docz'
import { Toggle } from './Toggle'

## Basic Usage

To implement the Toggle component, you need to import it first:

```js
import { Toggle } from '@react-ui-org/react-ui';
```

And use it:

<Playground>
  {() => {
    const [studioQuality, setStudioQuality] = React.useState(true);
    return (
      <Toggle
        changeHandler={() => setStudioQuality(!studioQuality)}
        checked={studioQuality}
        id="my-toggle"
        label="Listen in studio quality"
      />
    );
  }}
</Playground>

See [API](#api) for all available options.

## General Guidelines

- Use the toggle only for boolean (true/false) input **with an immediate effect
  (without confirmation)**. To toggle things on or off in forms that require to
  be submitted by a button, use rather the
  [CheckboxField](/components/ui/checkbox-field) component.

- **Use positive wording for the toggle label,** so that it's clear what will
  happen when the user turns on the toggle. Avoid negations such as “Don't send
  me any emails” which would mean that the user needs to turn the toggle **on**
  in order for something **not** to happen.

- Only make the Toggle's label invisible when there is **another visual clue**
  to guide users through using the input.

- When a short label is not enough, use **help texts to guide users** before
  they enter anything.

- Use **clear, calm error messages** when there's a problem with what they
  entered.

- The toggle is designed for **switching things on and off.** Don't use it to
  switch between two different things that cannot be described as on or off
  using a single label, eg. different viewing modes. In such cases, consider
  using the [ButtonGroup](/components/ui/button-group) component.

## Help Text

You may provide an additional help text to clarify what will be the impact of
turning the toggle on or off.

<Playground>
  {() => {
    const [studioQuality, setStudioQuality] = React.useState(true);
    return (
      <Toggle
        changeHandler={() => setStudioQuality(!studioQuality)}
        checked={studioQuality}
        helpText={
          'Unrivaled audio quality. Uses the MQA audio format. '
          + 'Transfers a lot of data.'
        }
        id="my-toggle-with-help-text"
        label="Listen in studio quality"
      />
    );
  }}
</Playground>

## Label Options

In some cases, it may be convenient to visually hide the toggle label. The label
remains accessible to assistive technologies.

<Playground>
  {() => {
    const [studioQuality, setStudioQuality] = React.useState(true);
    return (
      <Toggle
        changeHandler={() => setStudioQuality(!studioQuality)}
        checked={studioQuality}
        id="my-toggle-with-invisible-label"
        isLabelVisible={false}
        label="You cannot see this"
      />
    );
  }}
</Playground>

It's also possible to display the label before the input:

<Playground>
  {() => {
    const [studioQuality, setStudioQuality] = React.useState(true);
    return (
      <Toggle
        changeHandler={() => setStudioQuality(!studioQuality)}
        checked={studioQuality}
        id="my-toggle-with-label-before-input"
        label="Listen in studio quality"
        labelPosition="before"
      />
    );
  }}
</Playground>

## States

### Validation States

Validation states visually present the result of validation of the input. You
should always **provide a validation message for states other than valid** so
users know what happened and what action they should take or what options they
have.

<Playground>
  {() => {
    const [studioQuality, setStudioQuality] = React.useState(true);
    return (
      <>
        <Toggle
          changeHandler={() => setStudioQuality(!studioQuality)}
          checked={studioQuality}
          id="my-toggle-valid"
          label="Listen in studio quality"
          validationState="valid"
        />
        <Toggle
          changeHandler={() => setStudioQuality(!studioQuality)}
          checked={studioQuality}
          id="my-toggle-warning"
          label="Listen in studio quality"
          validationState="warning"
          validationText={
            'Looks like you are connected over cellular network. '
            + 'Please make sure you know what you are doing.'
          }
        />
        <Toggle
          changeHandler={() => setStudioQuality(!studioQuality)}
          checked={studioQuality}
          id="my-toggle-invalid"
          label="Listen in studio quality"
          validationState="invalid"
          validationText="Please upgrade your plan to make this option available."
        />
      </>
    );
  }}
</Playground>

### Disabled State

Disabled state makes the input unavailable.

<Playground>
  <Toggle label="Disabled toggle" id="my-toggle-disabled" disabled />
  <Toggle
    label="Disabled toggle, checked"
    id="my-toggle-checked-disabled"
    disabled
    checked
  />
</Playground>

## API

In addition to the options below, you can add any custom attributes that do not
interfere with the API, and they will be passed to the `input` HTML element.
This is useful mainly to improve component's accessibility.

<Props table of={Toggle} />

## Theming

Head to [Forms Theming](/customize/theming/forms) to see shared form theming
options. On top of that, the following options are available for Toggle.

| Custom Property                                                    | Description                                    |
|--------------------------------------------------------------------|------------------------------------------------|
| `--rui-FormField--check__input--toggle__width`             | Input width (height is shared with other check fields) |
| `--rui-FormField--check__input--toggle__border-radius`             | Input corner radius                            |
| `--rui-FormField--check__input--toggle__background-size`           | Input background size                          |
| `--rui-FormField--check__input--toggle--default__background-image` | Background image of unchecked input            |
| `--rui-FormField--check__input--toggle--default__background-position` | Background position of unchecked input      |
| `--rui-FormField--check__input--toggle--checked__background-image` | Background image of checked input              |
| `--rui-FormField--check__input--toggle--checked__background-position` | Background position of checked input        |
