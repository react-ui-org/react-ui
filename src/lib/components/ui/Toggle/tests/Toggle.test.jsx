import React from 'react';
import {
  shallow,
  mount,
} from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import sinon from 'sinon';
import { Toggle } from '../Toggle';

describe('rendering', () => {
  it('renders correctly mandatory props only', () => {
    const tree = shallow(<Toggle id="test" label="label" />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with all props', () => {
    const tree = shallow(<Toggle
      label="label"
      disabled
      id="test"
      value="value"
      description="some help"
      error="some error"
      isLabelVisible
      labelPosition="after"
      required
      checked
      validationState="warning"
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});

describe('functionality', () => {
  it('calls onChange() when checked', () => {
    const spy = sinon.spy();
    const component = mount(<Toggle
      id="test"
      label="label"
      changeHandler={spy}
    />);

    component
      .find('input')
      .simulate('change', {
        preventDefault: () => {},
        target: { value: 'on' },
      });
    expect(spy.calledOnce).toEqual(true);
  });
});
