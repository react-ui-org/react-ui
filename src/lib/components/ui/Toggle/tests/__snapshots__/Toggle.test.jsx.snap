// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly mandatory props only 1`] = `
<div
  className="root
        labelPositionAfter"
>
  <label
    className="inputWrap"
    htmlFor="test"
    id="test__label"
  >
    <input
      checked={false}
      className="input"
      disabled={false}
      id="test"
      name="test"
      onChange={null}
      required={false}
      type="checkbox"
    />
    <span
      className="label"
    >
      <span
        className="labelInner"
        id="test__labelText"
      >
        label
      </span>
    </span>
    <span
      className="toggle"
    />
  </label>
</div>
`;

exports[`rendering renders correctly with all props 1`] = `
<div
  className="root
        labelPositionAfter
        isRootStateWarning"
>
  <label
    className="inputWrap"
    htmlFor="test"
    id="test__label"
  >
    <input
      checked={true}
      className="input"
      disabled={true}
      id="test"
      name="test"
      onChange={null}
      required={true}
      type="checkbox"
      value="value"
    />
    <span
      className="label"
    >
      <span
        className="labelInner"
        id="test__labelText"
      >
        label
      </span>
    </span>
    <span
      className="toggle"
    />
  </label>
  <div
    className="description"
    id="test__descriptionText"
  >
    some help
  </div>
  <div
    className="error"
    id="test__errorText"
  >
    some error
  </div>
</div>
`;
